### PR TITLE
AV-183407: Fix: AVI shows the segment twice as it receives from 2 different CDVS.

### PR DESCRIPTION
**Pull request by akshayhavile**
Testing:
1. Done on setup with no duplicate network
  a. Create, update, delete L4 and Ingress 
  b. Create, update, delete L4 and Ingress with Aviinfrasetting
  
  Need to do:
1. Test on setup where duplicate networks are present.

### CODE REVIEW

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

File: internal/k8s/crdcontroller.go [hunk 1]
```diff
@@ -945,3 +945,32 @@ func addSeGroupLabel(key, segName string) {
 
 	avicache.ConfigureSeGroupLabels(clients.AviClient[aviClientLen], seGroup)
 }
+
+func SetAviInfrasettingVIPNetworks(netAviInfra []akov1alpha1.AviInfraSettingVipNetwork) {
+	// assign the last avi client for ref checks
+	clients := avicache.SharedAVIClients()
+	aviClientLen := lib.GetshardSize()
+
+	network := avicache.PopulateVipNetworkwithUUID(clients.AviClient[aviClientLen], netAviInfra)
```
Comment (aaha97) [inline internal/k8s/crdcontroller.go:954]: should we add a cloud type check here?

**Comment by akshayhavile** [comment]
build ako

File: buildsettings.json [hunk 2]
```diff
@@ -1,11 +1,11 @@
 {
-    "version": "1.10.1",
+    "version": "1.10.2",
     "avi": {
         "maxVersion": "22.1.4",
         "minVersion": "21.1.5"
     },
     "kubernetes": {
-        "maxVersion": "1.26",
+        "maxVersion": "1.27",
```
Comment (swathinsankaran) [inline buildsettings.json:8]: Do we have to claim the new K8s version in a patch release?

File: internal/cache/controller_obj_cache.go [hunk 3]
```diff
@@ -32,6 +32,7 @@ import (
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/third_party/github.com/vmware/alb-sdk/go/clients"
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/third_party/github.com/vmware/alb-sdk/go/session"
 
+	pq "github.com/jupp0r/go-priority-queue"
```
Comment (swathinsankaran) [inline internal/cache/controller_obj_cache.go:35]: Should we go with a pq implementation with more forks and stars such as https://github.com/oleiade/lane which has generics support?

**Review by aaha97 — APPROVED** [review]
LGTM

File: buildsettings.json (hunk 2 above)
Comment (swathinsankaran) [reply buildsettings.json:8]: Got it. This is an ask from the TKG team.

File: pkg/apis/ako/v1alpha1/aviinfrasetting.go [hunk 4]
```diff
@@ -55,12 +55,14 @@ type AviInfraListeners struct {
 }
 type AviInfraSettingVipNetwork struct {
 	NetworkName string `json:"networkName,omitempty"`
+	NetworkUUID string `json:"networkUUID,omitempty"`
```
Comment (swathinsankaran) [inline pkg/apis/ako/v1alpha1/aviinfrasetting.go:58]: The crd yaml changes corresponding to this seem to be missing.

File: internal/cache/controller_obj_cache.go [hunk 5]
```diff
@@ -2597,7 +2598,118 @@ func ValidateUserInput(client *clients.AviClient) (bool, error) {
...
+
+// Function to find max host overlap between mgmt network and vipNetwork/nodenetwoeklist
+func findHostWithMaxOverlapping(client *clients.AviClient, localNetworkList []models.Network) akov1alpha1.AviInfraSettingVipNetwork {
+	cloudMgmtNW := lib.GetCloudMgmtNetwork()
+	//maxLenMatch := 0
+	var matchedNW akov1alpha1.AviInfraSettingVipNetwork
+	mgmtHostRefs := findHostRefs(client, cloudMgmtNW)
+	mgmtHostsSet := sets.NewString(mgmtHostRefs...)
+	//default choice of network
+	desiredNW := localNetworkList[0]
```
Comment (akshayhavile) [inline internal/cache/controller_obj_cache.go:2624]: change default to maximum host overlapping entry from priority queue.

File: internal/cache/controller_obj_cache.go [hunk 6]
```diff
@@ -2956,16 +3075,26 @@ func checkIPAMForUsableNetworkLabels(client *clients.AviClient, ipamRefUri *stri
...
+		}
+		lib.SetVipNetworkList(vipListUpdated)
 		return true, nil
 	}
 
 	// 2. AKO created VIP network for AKO in VCF
 	if utils.IsVCFCluster() {
-		vipNetList := akov1alpha1.AviInfraSettingVipNetwork{
-			NetworkName: lib.GetVCFNetworkName(),
+		vipNetList := []akov1alpha1.AviInfraSettingVipNetwork{
```
Comment (akshayhavile) [inline internal/cache/controller_obj_cache.go:3088]: remove this code.

File: internal/cache/controller_obj_cache.go [hunk 7]
```diff
@@ -2998,9 +3127,14 @@ func checkIPAMForUsableNetworkLabels(client *clients.AviClient, ipamRefUri *stri
 		}
 
 		if markerNetworkFound != "" {
-			lib.SetVipNetworkList([]akov1alpha1.AviInfraSettingVipNetwork{{
+			vipList := []akov1alpha1.AviInfraSettingVipNetwork{{
```
Comment (akshayhavile) [inline internal/cache/controller_obj_cache.go:3130]: remove this code.

File: internal/k8s/crdcontroller.go [hunk 8]
```diff
@@ -945,3 +945,36 @@ func addSeGroupLabel(key, segName string) {
...
+	var err error
+
+	for _, net := range netAviInfra {
+		nwMap := lib.NodeNetworkMap{
+			Cidrs: net.Cidrs,
+		}
+		nodeNetorkList[net.NetworkName] = nwMap
+	}
+
+	avicache.FetchNodeNetworks(clients.AviClient[aviClientLen], &err, nodeNetorkList)
```
Comment (akshayhavile) [inline internal/k8s/crdcontroller.go:976]: Add cloudtype check here.

File: internal/cache/controller_obj_cache.go (hunk 5 above)
Comment (akshayhavile) [reply internal/cache/controller_obj_cache.go:2624]: done

File: internal/cache/controller_obj_cache.go (hunk 6 above)
Comment (akshayhavile) [reply internal/cache/controller_obj_cache.go:3154]: done

**Comment by akshayhavile** [comment]
build ako

File: internal/cache/controller_obj_cache.go [hunk 9]
```diff
@@ -2597,7 +2598,184 @@ func ValidateUserInput(client *clients.AviClient) (bool, error) {
...
+		if networkFound {
+			// If cidr matched network found, reset all list and add only that network
+			localVIPNetwork = network
+			break
+		}
+	}
+	return networkFound, localVIPNetwork
+}
+
+func PopulateVipNetworkwithUUID(client *clients.AviClient, vipNetworks []akov1alpha1.AviInfraSettingVipNetwork) []akov1alpha1.AviInfraSettingVipNetwork {
```
Comment (swathinsankaran) [inline internal/cache/controller_obj_cache.go:2708]: Should we test this with a non-admin user who has limited permissions?

**Review by swathinsankaran — APPROVED** [review]
Overall changes look good to me.

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Pull request merged**